### PR TITLE
Fix test classes to work with segrep.

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR fixes some test classes to work with segrep

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java [hunk 1]
```diff
@@ -187,21 +189,23 @@ public void testExplainWithSource() throws Exception {
...
+            assertTrue(response.getExplanation().isMatch());
+            assertThat(response.getExplanation().getValue(), equalTo(1.0f));
+            assertThat(response.getGetResult().isExists(), equalTo(true));
+            assertThat(response.getGetResult().getId(), equalTo("1"));
+            assertThat(response.getGetResult().getSource().size(), equalTo(1));
+            assertThat(((Map<String, Object>) response.getGetResult().getSource().get("obj1")).get("field1").toString(), equalTo("value1"));
+        });
 
-        response = client().prepareExplain(indexOrAlias(), "1")
+        ExplainResponse response = client().prepareExplain(indexOrAlias(), "1")
```
Comment (Poojita-Raj) [inline server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java:208]: why was this change needed? response in line 68 has already been reused once on line 87.

File: server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java [hunk 2]
```diff
@@ -267,9 +271,17 @@ public void testExplainDateRangeInQueryString() {
...
+        try {
+            assertBusy(() -> {
+                ExplainResponse explainResponse = client().prepareExplain("test", "1")
+                    .setQuery(queryStringQuery("past:[now-2M/d TO now/d]"))
+                    .get();
+                assertThat(explainResponse.isExists(), equalTo(true));
+                assertThat(explainResponse.isMatch(), equalTo(true));
+            });
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (Poojita-Raj) [inline server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java:283]: I think we can throw an Opensearch exception?

File: server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java [hunk 3]
```diff
@@ -652,122 +660,124 @@ public void testSimpleStats() throws Exception {
...
+            // check get
+            GetResponse getResponse = client().prepareGet("test2", "1").execute().actionGet();
+            assertThat(getResponse.isExists(), equalTo(true));
 
-        stats = client().admin().indices().prepareStats().execute().actionGet();
-        assertThat(stats.getTotal().getGet().getCount(), equalTo(2L));
-        assertThat(stats.getTotal().getGet().getExistsCount(), equalTo(1L));
-        assertThat(stats.getTotal().getGet().getMissingCount(), equalTo(1L));
+            stats = client().admin().indices().prepareStats().execute().actionGet();
+            assertThat(stats.getTotal().getGet().getCount(), equalTo(1L));
```
Comment (Poojita-Raj) [inline server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java:712]: why are we changing values of the count?

Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java:712]: looks like they got in by mistake. I will update this

File: server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java (hunk 2 above)
Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java:283]: Sure, I will update this.

Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/explain/ExplainActionIT.java:283]: You can add throws clause to method definition similar to what we have done for other methods.

File: server/src/internalClusterTest/java/org/opensearch/index/store/ExceptionRetryIT.java [hunk 4]
```diff
@@ -140,32 +140,38 @@ public void testRetryDueToExceptionOnNetworkLayer() throws ExecutionException, I
...
-                    logger.info("will not print anymore in case more duplicates are found.");
-                    found_duplicate_already = true;
                 }
-                dupCounter++;
-            }
+                assertSearchResponse(searchResponse);
+                assertThat(dupCounter, equalTo(0L));
+                assertHitCount(searchResponse, numDocs);
+            });
+        } catch (Exception e) {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/index/store/ExceptionRetryIT.java:172]: You can add `Exception` in the method definition throws clause and let test runtime handle it.

File: server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java [hunk 5]
```diff
@@ -822,17 +832,23 @@ public void testSegmentsStats() {
...
+
+                client().admin().indices().prepareFlush().get();
+                client().admin().indices().prepareForceMerge().setMaxNumSegments(1).execute().actionGet();
+                client().admin().indices().prepareRefresh().get();
+                stats = client().admin().indices().prepareStats().setSegments(true).get();
+
+                assertThat(stats.getTotal().getSegments(), notNullValue());
+                assertThat(stats.getTotal().getSegments().getCount(), equalTo((long) test1.totalNumShards));
+            });
+        } catch (Exception e) {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java:849]: Same as above. We can add exception to method signature throws clause.

File: server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java [hunk 6]
```diff
@@ -156,8 +162,14 @@ public void testSimpleSearchRouting() {
...
         logger.info("--> verifying get with routing, should find");
-        for (int i = 0; i < 5; i++) {
-            assertThat(client().prepareGet("test", "1").setRouting(routingValue).execute().actionGet().isExists(), equalTo(true));
+        try {
+            assertBusy(() -> {
+                for (int i = 0; i < 5; i++) {
+                    assertThat(client().prepareGet("test", "1").setRouting(routingValue).execute().actionGet().isExists(), equalTo(true));
+                }
+            });
+        } catch (Exception e) {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java:171]: Same as above.

File: server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java [hunk 7]
```diff
@@ -228,21 +240,32 @@ public void testSimpleSearchRouting() {
...
+                            .setQuery(QueryBuilders.matchAllQuery())
+                            .execute()
+                            .actionGet()
+                            .getHits()
+                            .getTotalHits().value,
+                        equalTo(2L)
+                    );
+                }
+            });
+        } catch (Exception e) {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java:267]: Same as above

File: server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/TermsDocCountErrorIT.java [hunk 8]
```diff
@@ -276,25 +276,31 @@ private void assertNoDocCountErrorSingleResponse(int size, SearchResponse testRe
...
+
+                for (Terms.Bucket testBucket : testBuckets) {
+                    assertThat(testBucket, notNullValue());
+                    Terms.Bucket accurateBucket = accurateTerms.getBucketByKey(testBucket.getKeyAsString());
+                    assertThat(accurateBucket, notNullValue());
+                    assertThat(accurateBucket.getDocCountError(), equalTo(0L));
+                    assertThat(testBucket.getDocCountError(), anyOf(equalTo(-1L), equalTo(0L)));
+                }
+            });
+        } catch (Exception e) {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/TermsDocCountErrorIT.java:302]: same as above.

File: test/framework/src/main/java/org/opensearch/repositories/blobstore/OpenSearchBlobStoreRepositoryIntegTestCase.java [hunk 9]
```diff
@@ -344,9 +345,11 @@ public void testSnapshotAndRestore() throws Exception {
         // higher timeout since we can have quite a few shards and a little more data here
         ensureGreen(TimeValue.timeValueSeconds(120));
 
-        for (int i = 0; i < indexCount; i++) {
-            assertHitCount(client().prepareSearch(indexNames[i]).setSize(0).get(), docCounts[i]);
-        }
+        assertBusy(() -> {
+            for (int i = 0; i < indexCount; i++) {
+                assertHitCount(client().prepareSearch(indexNames[i]).setSize(0).get(), docCounts[i]);
```
Comment (dreamer-89) [inline test/framework/src/main/java/org/opensearch/repositories/blobstore/OpenSearchBlobStoreRepositoryIntegTestCase.java:350]: Do we need assertBusy on `assertHitCount` statement ?

File: server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java [hunk 10]
```diff
@@ -375,9 +398,14 @@ public void testRequiredRoutingCrudApis() throws Exception {
         }
 
         logger.info("--> verifying get with routing, should find");
-        for (int i = 0; i < 5; i++) {
-            assertThat(client().prepareGet(indexOrAlias(), "1").setRouting(routingValue).execute().actionGet().isExists(), equalTo(true));
-        }
+        assertBusy(() -> {
```
Comment (dreamer-89) [inline server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java:401]: Using assertBusy will re-execute the code block inside on assertion trip. One single assertion failure, get requests will start from 0, which is probably not needed. This is fine for code block which doesn't execute API requests. We can use assertBusy on inner assertion statement. This is applicable for this logic and other places where we asserted on entire code block.

File: test/framework/src/main/java/org/opensearch/repositories/blobstore/OpenSearchBlobStoreRepositoryIntegTestCase.java (hunk 9 above)
Comment (Rishikesh1159) [reply test/framework/src/main/java/org/opensearch/repositories/blobstore/OpenSearchBlobStoreRepositoryIntegTestCase.java:350]: yes, as assertHitCount is making an API call, it's response might change in a assert busy.

I have updated it to something like this :
```
for (int i = 0; i < indexCount; i++) {
            final int j = i;
            assertBusy(() -> {
                assertHitCount(client().prepareSearch(indexNames[j]).setSize(0).get(), docCounts[j]);
            });
        }
```

File: server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java (hunk 10 above)
Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/routing/SimpleRoutingIT.java:401]: Got it. Will update this

File: server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java (hunk 3 above)
Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/indices/stats/IndexStatsIT.java:712]: Looks like I haven't changed these, the git diff shows like values have been changed, but actually they haven't. It was just refactoring same code blocks up and down. The git diff makes the changes a little confusing, but no values of count are changed here.

**Review by dreamer-89 — APPROVED** [review]
